packages/discovery-provider/src/queries/query_helpers.py

```diff
@@ -545,7 +557,130 @@ def populate_track_metadata(
+            )
+            if stream_signature:
+                return get_content_url_with_mirrors(
+                    stream_signature,
+                    track["track_cid"],
+                    track.get("placement_hosts", None),
+                )
+    return {"url": None, "mirrors": []}
+
+
+def get_preview_url_with_mirrors(track: dict, user_id: int | None) -> UrlWithMirrors:
+    logger.error(f"preview_cid={track.get('preview_cid')}")
+    if track.get("preview_cid"):
+        preview_signature = get_gated_content_signature(
+            {
+                "track_id": track["track_id"],
+                "cid": track["preview_cid"],
+                "type": "track",
+                "user_id": user_id,
+                "is_gated": False,
+            }
+        )
+        if preview_signature:
+            return get_content_url_with_mirrors(
+                preview_signature,
+                track["preview_cid"],
+                track.get("placement_hosts", None),
+            )
+    return {"url": None, "mirrors": []}
+
+
+def get_download_url_with_mirrors(
+    track: dict, user_id: int | None, is_authorized_as_user: bool
+) -> UrlWithMirrors:
+    # If the cid exists and the user has access
+    if track.get("orig_file_cid", False) and track.get("access", {}).get(
+        "download", False
+    ):
+        # If the signature is authorized for the user that has access or if there's no access gate
+        if is_authorized_as_user or not track.get("is_download_gated", False):
+            download_signature = get_gated_content_signature(
+                {
+                    "track_id": track["track_id"],
+                    "cid": track["orig_file_cid"],
+                    "type": "track",
+                    "user_id": user_id,
+                    "is_gated": track.get("is_download_gated", False),
+                }
+            )
+            if download_signature:
+                return get_content_url_with_mirrors(
+                    download_signature,
+                    track["orig_file_cid"],
+                    track.get("placement_hosts", None),
+                )
+    return {"url": None, "mirrors": []}
+
+
+def get_content_url_with_mirrors(
+    signature: GatedContentSignature, cid: str, placement_hosts=None
+) -> UrlWithMirrors:
+    params = {"signature": json.dumps(signature)}
+
+    base_path = f"tracks/cidstream/{cid}"
+    query_string = urlencode(params, quote_via=quote)
+    path = f"{base_path}?{query_string}"
+
+    content_nodes = []
+    # if track has placement_hosts, use that instead
+    if placement_hosts:
+        content_nodes = placement_hosts.split(",")
+    else:
+        healthy_nodes = get_all_healthy_content_nodes_cached(redis)
+        if healthy_nodes:
+            rendezvous = RendezvousHash(
+                *[re.sub("/$", "", node["endpoint"].lower()) for node in healthy_nodes]
+            )
+
+            content_nodes = rendezvous.get_n(REPLICATION_FACTOR, cid)
```

Comment:
not sure how expensive this is. definitely better than what we currently do client side, but possible future optimization

Reply:
gotta do it at some point 🤷  - tradeoff is in responsiveness of the play button vs in the initial load/render